src/yubikey.rs

```diff
@@ -357,37 +357,40 @@ impl YubiKey {
     }
 
     /// Authenticate to the card using the provided management key (MGM).
-    pub fn authenticate(&mut self, mgm_key: MgmKey) -> Result<()> {
+    pub fn authenticate<C: MgmKeyAlgorithm>(&mut self, mgm_key: MgmKey<C>) -> Result<()> {
         let txn = self.begin_transaction()?;
 
         // get a challenge from the card
-        let challenge = Apdu::new(Ins::Authenticate)
-            .params(ALGO_3DES, KEY_CARDMGM)
+        let card_response = Apdu::new(Ins::Authenticate)
+            .params(mgm_key.algorithm_id(), KEY_CARDMGM)
             .data([TAG_DYN_AUTH, 0x02, 0x80, 0x00])
             .transmit(&txn, 261)?;
 
-        if !challenge.is_success() || challenge.data().len() < 12 {
+        if !card_response.is_success() {
+            dbg!("here-auth");
             return Err(Error::AuthenticationError);
         }
 
         // send a response to the cards challenge and a challenge of our own.
-        let response = mgm_key.decrypt(challenge.data()[4..12].try_into()?);
+        let card_challenge = mgm_key.card_challenge(&card_response.data()[4..])?;
```

Comment:
This was a bit tricky and the documentation of the APDU's is not clear.

I would suggest that what is to be sent through AES would be 8 bytes, however that wouldnt match the block size and I dont think would work.

Peeking at the `ykman` C code, it appears that in there they use the block length of the cipher being used 8 bytes for 3des due to being 56bit and 16 bytes due to AES's block being 128bits.

As such I modified this to use sizes from the challenge instead.

If this is unclear I can add some documentation to this effect, or we can talk to Yubico about a minor issue in their documentation. WDYT?

Reply:
Maybe a pointer to the ykman C code here (a github permalink, perhaps) would be nice here.

Reply:
I think what you're looking for is this bit:
https://github.com/Yubico/yubikey-manager/blob/b6d1bfdb6aa580beb26de61f6393be879a2e5f23/yubikit/piv.py#L160-L163

So 8 bytes for three DES, and 16 bytes for AES.

I've otherwise found yubico very responsive on PRs on their documentation,